### ISSUE TITLE
Dialogs switch the stack to an incorrect window

### ISSUE THREAD

**Issue by sitepodmatt** [pop-os/shell#712]
(1) Issue/Bug Description
When using window stacking, popos changes active stack window when a IntelliJ IDEA based product in a stack application launches a dialog (such as Settings)

(2) Steps to reproduce (if you know):

Create a popos stack
Add a Jetbrain's IntelliJ IDEA (could also be a variant such as WebStorm, Rider, CLion, PyCharm, PhpStorm or RubyMine, DataGrip, AppCode or Android Studio)
Add another app such as Firefox.  
Do an action that launches a dialog from  the IntelliJ such as opening settings (File -> Settings) popos shell needlessly switches the stack to the wrong window. 

(3) Expected behavior:  No switching of my stack

(4) Distribution (run cat /etc/os-release): Pop!_OS 20.10

(5) Gnome Shell version: 3.38.1

(6) Pop Shell version (run apt policy pop-shell or provide the latest commit if building locally):

1.0.0~1604688242~20.10~d1d2b0e

(7) Where was Pop Shell installed from: apt-get install popos-desktop

I found this https://github.com/pop-os/shell/issues/407 but it doesn't explain why the stack is changing active window, unless it's expecting the dialog to open as part of the stack and trying to focus it but it isn't there?? 

Perhaps it should only refocus the stack if a window is actually added to the stack, as the modal appears correctly in all other manners?

**Comment by sitepodmatt** [comment]
![image](https://user-images.githubusercontent.com/1687145/98964537-363f8d00-2509-11eb-9e7c-ae75e14f3ab2.png)

tl;dr popos is changing window the active window in the stack when it shouldnt

**Comment by sitepodmatt** [comment]
NAME="Pop!_OS"
VERSION="20.10"
ID=pop
ID_LIKE="ubuntu debian"
PRETTY_NAME="Pop!_OS 20.10"
VERSION_ID="20.10"
HOME_URL="https://pop.system76.com"
SUPPORT_URL="https://support.system76.com"
BUG_REPORT_URL="https://github.com/pop-os/pop/issues"
PRIVACY_POLICY_URL="https://system76.com/privacy"
VERSION_CODENAME=groovy
UBUNTU_CODENAME=groovy
LOGO=distributor-logo-pop-os

**Comment by mmstick** [comment]
Dialogs aren't permitted to tile, so something else is happening. I've been looking into this

**Comment by sitepodmatt** [comment]
Thanks. I've updated the bug description tl;dr - easier to replicate via just opening settings rather than trying to get a rename window going.

**Comment by sitepodmatt** [comment]
Was this fixed by https://github.com/pop-os/shell/commit/2a3560ac4b273225a67c939933536655a50ab14c?

**Comment by sitepodmatt** [comment]
Just tried with latest verison, still busted, in stacking mode, opening a dialog like rename variable (shift+f6) switches the stack window

**Comment by sitepodmatt** [comment]
Just to update, renaming has become inline in many Jetbrains IDEs, ill choose most popular IntelliJ,  so stacking setup with three windows (ff, intellij, terminal), so easiest way now is to shift twice in quick succession (brings up the All tasks selector), and choose the first option for example (Analyse Stack trace or thread dump), pop os gets confused and focuses the nth window in the stack (the terminal) whilst this dialog is showing, and back to nth-1 (the middle - intellij ) after closing..

**Comment by sitepodmatt** [comment]
To be clear, this is not a problem in i3wm on the same popos distro (and thus same intellij), and seems to only happen in stacking mode

**Comment by hoshsadiq** [comment]
Hello all, I have the same issue. I don't know much about the inner workings here but it looks to me like an off by one index somewhere. It seems to always switch to the last window I have selected, i.e. if I Firefox, Spotify, IntelliJ, if I open Firefox, then IntelliJ, then open a dialog, it shows Firefox, however, if I open Spotify, then IntelliJ, dialogs switch to Spotify.

Also, dragging editor windows also causes this, which makes dropping windows extremely hard.

**Comment by sitepodmatt** [comment]
Yes, yeah I initially thought it off by one, I think i's expecting the dialog to be part of the stack and then it tries to switch to it. 

@mmstick would any diagnosis material help on this? a video via dropbox for example?

**Comment by sitepodmatt** [comment]
Think this maybe related to the dialogs having _NET_WM_STATE_SKIP_TASKBAR

**Comment by sitepodmatt** [comment]
This is still happening in the latest dev branch, although not as much. 

activate is getting called in stack.ts, and then usually the first window is getting selected due to Ecs.entity_eq(entity, component.entity) returning true. There are times I can't replicate this though now until I select a different window in stack and then switch back to IntelliJ then it's replicatable again. Will try and do some more exhuastive debugging

**Comment by sitepodmatt** [comment]
getting closer get_meta_window().is_skip_taskbar() is true for this dialog...

**Comment by sitepodmatt** [comment]
1616834689767=== activate called == 
ist :false
ist :activate@/home/matt/.local/share/gnome-shell/extensions/pop-shell@system76.com/stack.js:79:31
on_focused@/home/matt/.local/share/gnome-shell/extensions/pop-shell@system76.com/extension.js:490:197
signals_attach/</<@/home/matt/.local/share/gnome-shell/extensions/pop-shell@system76.com/extension.js:1319:26

ist2:false
1616834689779=== activate called == 
ist :false
ist :activate@/home/matt/.local/share/gnome-shell/extensions/pop-shell@system76.com/stack.js:79:31
on_focused@/home/matt/.local/share/gnome-shell/extensions/pop-shell@system76.com/extension.js:490:197
signals_attach/</<@/home/matt/.local/share/gnome-shell/extensions/pop-shell@system76.com/extension.js:1319:26

Here how activate is being called... seem to come from focus

**Comment by sitepodmatt** [comment]
```
    focus_window(): Window.ShellWindow | null {
        let focused = this.get_window(display.get_focus_window())
        //if (!focused && this.last_focused) {
        //    focused = this.windows.get(this.last_focused);
        //}
        return focused;
    }

```
Okay managed to get this solved for my use case, but unsure what other problems I've introduced here by simply commenting out.

**Comment by sitepodmatt** [comment]
Revising src/extensions.ts:L590
```
    focus_window(): Window.ShellWindow | null {
        let focused = this.get_window(display.get_focus_window())
        if (!focused && this.last_focused ) {
            focused = this.windows.get(this.last_focused);
            if (focused && focused.stack !== null) {
              focused = null
            }
        }
        return focused;
    }
```
This now works and I've reduce the hack to apply to stack only, I'm heavily skeptical that this is the correct fix though, @mmstick any thoughts on further troubleshooting this?

**Comment by sitepodmatt** [comment]
it seems this can also be solved by wrapping the 'notify::focus-window' with a GLib.timeout_add 300ms,  as get_focus_window() does not return null after 300ms, perhasp suggesting Swing dialogs are somewhat delayed by coming into focus after whatever event notifies new dialog. Interesting..

**Comment by mmstick** [comment]
300ms delay breaks active hint and management mode

**Comment by sitepodmatt** [comment]
Eek. I wonder if we changed it to fire immediately if there is a focused window, and only if falsey, then schedule a timeout for 300ms to run?

**Comment by hoshsadiq** [comment]
I've just upgraded my machine's packages (pop-shell==`1.1.0~1620418263~20.10~9507dc3`) and the behaviour has gotten worse. When any of the previous popups show up in IntelliJ it completely hides IntelliJ and the popup. It is impossible to change back to IntelliJ unless I ``Alt+` `` twice which brings back the popup, but not the main IntelliJ window.

**Comment by lhotari** [comment]
I upgraded to Pop OS 21.04 beta & COSMIC. This problem exists and makes window stacking unusable with running IntelliJ or other Jetbrains IDEs. 

@mmstick is there a plan to address this issue?

**Comment by mmstick** [comment]
@lhotari All issues are planned to be addressed, but wanting to fix them doesn't automatically make the solution known.

**Comment by hoshsadiq** [comment]
Is there any way to permanently disable the stacks feature until this gets fixed? IntelliJ is pretty much unusable with this bug. I feel like I've seen the option previously but I can't find it any more.

**Comment by lhotari** [comment]
> All issues are planned to be addressed, but wanting to fix them doesn't automatically make the solution known.

@mmstick I really love Pop OS Shell. Pop OS Shell has the best window manager that I have ever used and the stacking and auto-tiling features are incredible. This single bug makes the user experience terrible. The reason is that I use IntelliJ and this bug shows up mainly with IntelliJ (and other Jetbrains IDE products).
If there was a bug bounty, I'd pay cash to get this fixed. :)

**Comment by lhotari** [comment]
The problem might be caused by IntelliJ. I found this blog post: https://makandracards.com/makandra/40343-how-to-fix-rubymine-intellij-find-file-dialog-losing-focus-on-awesome-wm . It suggests enabling `focus.follows.mouse.workarounds` registry setting in IntelliJ. It references https://youtrack.jetbrains.com/issue/IDEA-112015 . I tried this and also the other setting `ide.settings.move.mouse.on.default.button`. but enabling those settings doesn't make a difference.

**Comment by lhotari** [comment]
I found a workaround for using IntelliJ. If I change the focus to some other window outside of the stack and then switch the focus back to IntelliJ, the dialogs get focus as expected. The problem with IntelliJ seems to occur when another window such as Gnome Terminal or Chrome has been the previous window as part of the same stack. Switching between multiple IntelliJ windows in the same stack doesn't cause the problem.

**Comment by lhotari** [comment]
Even though the [blog post "How to fix: RubyMine / IntelliJ "find file" dialog losing focus on awesome wm"](https://makandracards.com/makandra/40343-how-to-fix-rubymine-intellij-find-file-dialog-losing-focus-on-awesome-wm) is in different context, I wonder if the details in about using `sun-awt-X11-XDialogPeer` and `sun-awt-X11-XFramePeer` for detecting the focused dialog would be helpful for fixing this issue?

This was the lua script in 
```lua
-- Enable sloppy focus (without defocusing IntelliJ dialogs)
c:connect_signal("mouse::enter", function(c)
    local focused = client.focus
    if focused 
        and focused.class == c.class
        and focused.instance == "sun-awt-X11-XDialogPeer"
        and c.instance == "sun-awt-X11-XFramePeer"
        then
            return
    end

    if awful.layout.get(c.screen) ~= awful.layout.suit.magnifier
        and awful.client.focus.filter(c) then
        client.focus = c
    end
end)
```

something similar can be found in i3 workarounds for Java applications: https://stackoverflow.com/a/32596391/ 

```
for_window [instance="sun-awt-X11-XFramePeer"] floating enable
for_window [instance="sun-awt-X11-XDialogPeer"] floating enable
```

also found in GitHub here "Kludge IntelliJ IDEA splash popup to be transient":
https://github.com/raboof/notion/commit/995bca3ad61a8e33b7828b5a426db5b58f7cc82c
That seems to be kludges for [Notion tiling, tabbed window manager for the X window system](https://notionwm.net/).

Perhaps we need some kludges for IntelliJ in Pop-OS Shell?

**Comment by lhotari** [comment]
I used this type of shell function to get information about the popup dialog:
```bash
function show_info_for_new_jetbrains_windows() {
    ws=$(xdotool search --onlyvisible --class 'jetbrains.*')
    while true; do
        wstmp=$(xdotool search --onlyvisible --class 'jetbrains.*')
        wsnew=$(diff <(echo "$ws") <(echo "$wstmp") | awk '/^>/ {print $2}')
        echo found windows: $wsnew
        for win in $wsnew; do
            xwininfo -id $win -all
            xprop -id $win
        done
        ws=$wstmp
        sleep 0.5
    done
}
```
(function based on script in https://github.com/raboof/notion/issues/321#issuecomment-798447191)

Here's the output for the "Find in Files" popup dialog in IntelliJ Idea Ultimate:

```
found windows: 96472735
96472738

xwininfo: Window id: 0x5c00e9f " "

  Root window id: 0x7a8 (the root window) (has no name)
  Parent window id: 0x7a8 (the root window) (has no name)
     2 children:
     0x5c00ea5 "FocusProxy": ("Focus-Proxy-Window" "FocusProxy")  1x1+-1+-1  +2012+1105
     0x5c00ea2 "Content window": ("jetbrains-idea" "jetbrains-idea")  1357x758+0+0  +2013+1106

  Absolute upper-left X:  2013
  Absolute upper-left Y:  1106
  Relative upper-left X:  2013
  Relative upper-left Y:  1106
  Width: 1357
  Height: 758
  Depth: 24
  Visual: 0x21
  Visual Class: TrueColor
  Border width: 0
  Class: InputOutput
  Colormap: 0x20 (installed)
  Bit Gravity State: NorthWestGravity
  Window Gravity State: NorthWestGravity
  Backing Store State: NotUseful
  Save Under State: no
  Map State: IsViewable
  Override Redirect State: no
  Corners:  +2013+1106  -470+1106  -470-296  +2013-296
  -geometry 1357x758+2013+1106

  Bit gravity: NorthWestGravity
  Window gravity: NorthWestGravity
  Backing-store hint: NotUseful
  Backing-planes to be preserved: 0xffffffff
  Backing pixel: 0
  Save-unders: No

  Someone wants these events:
      ButtonPress
      ButtonRelease
      EnterWindow
      LeaveWindow
      PointerMotion
      ButtonMotion
      Exposure
      VisibilityChange
      StructureNotify
      PropertyChange
      OwnerGrabButton
  Do not propagate these events:
  Override redirection?: No

  Window manager hints:
      Client accepts input or input focus: No
      Displayed on desktop 0
      Window type:
          Dialog
      Window state:
          Skip Taskbar
          Focused
      Process id: 313688 on host pop-os
      Frame extents: 0, 0, 37, 0

  Normal window size hints:
      User supplied location: 2013, 1106
      Program supplied location: 2013, 1106
      Program supplied size: 1357 by 758
      Program supplied minimum size: 535 by 256
      Program supplied window gravity: NorthWestGravity
  No zoom window size hints defined

  No window shape defined
  No border shape defined

WM_STATE(WM_STATE):
		window state: Normal
		icon window: 0x0
_NET_WM_DESKTOP(CARDINAL) = 0
_GTK_EDGE_CONSTRAINTS(CARDINAL) = 170
_NET_WM_ALLOWED_ACTIONS(ATOM) = _NET_WM_ACTION_MOVE, _NET_WM_ACTION_RESIZE, _NET_WM_ACTION_CHANGE_DESKTOP, _NET_WM_ACTION_CLOSE, _NET_WM_ACTION_ABOVE, _NET_WM_ACTION_BELOW
_NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_DIALOG
_OL_DECOR_DEL(ATOM) = _OL_DECOR_HEADER, _OL_DECOR_RESIZE, _OL_DECOR_CLOSE
_MOTIF_WM_HINTS(_MOTIF_WM_HINTS) = 0x3, 0x26, 0x0, 0x0, 0x0
_NET_FRAME_EXTENTS(CARDINAL) = 0, 0, 37, 0
_MOTIF_DRAG_RECEIVER_INFO(_MOTIF_DRAG_RECEIVER_INFO) = 0x6c, 0x0, 0x5, 0x0, 0x9f, 0xe, 0xc0, 0x5, 0x0, 0x0, 0x0, 0x0, 0x10, 0x0, 0x0, 0x0
XdndAware(ATOM) = BITMAP
_NET_WM_STATE(ATOM) = _NET_WM_STATE_SKIP_TASKBAR, _NET_WM_STATE_FOCUSED
WM_HINTS(WM_HINTS):
		Client accepts input or input focus: False
		window id # of group leader: 0x5c0006c
WM_TRANSIENT_FOR(WINDOW): window id # 0x5c0006c
_NET_WM_PID(CARDINAL) = 313688
WM_CLIENT_MACHINE(STRING) = "pop-os"
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, WM_TAKE_FOCUS
WM_CLASS(STRING) = "jetbrains-idea", "jetbrains-idea"
_NET_WM_USER_TIME_WINDOW(WINDOW): window id # 0x5c00008
WM_CLIENT_LEADER(WINDOW): window id # 0x5c00008
_NET_WM_ICON_NAME(UTF8_STRING) = "Java"
WM_ICON_NAME(STRING) = "Java"
_NET_WM_NAME(UTF8_STRING) = " "
WM_NAME(STRING) = " "
WM_NORMAL_HINTS(WM_SIZE_HINTS):
		user specified location: 2013, 1106
		program specified location: 2013, 1106
		program specified size: 1357 by 758
		program specified minimum size: 535 by 256
		window gravity: NorthWest
```

**Comment by lhotari** [comment]
Switching to the wrong active window seems to be caused by this piece of code:
https://github.com/pop-os/shell/blob/master/src/extension.ts/#L629-L632

This helped find another workaround:
https://github.com/pop-os/shell/blob/master/src/extension.ts/#L720-L724

`last_focused` state gets cleared by switching the workspace. That's the workaround I'll use for IntelliJ until the issue gets fixed.

**Comment by lhotari** [comment]
@mmstick I have created a PR that fixes this issue. Please review #1007 . Thanks to @sitepodmatt for pointing out the root cause of the issue.

**Comment by hoshsadiq** [comment]
Thanks @lhotari, @mmstick. That PR has fixed it the issue for me.
